File: frontend/sentinel/src/app/core/services/incident.service.ts
import { HttpClient, HttpParams } from '@angular/common/http';
import { Injectable, inject, signal } from '@angular/core';
import { Observable, catchError, map, of, tap } from 'rxjs';
import { environment } from '../../../environments/environment';
import {
  Incident,
  IncidentActionType,
  IncidentNote,
  IncidentStatus,
  ThreatHuntingFilter,
  ThreatHuntingResult
} from '../models/incident.model';

@Injectable({ providedIn: 'root' })
export class IncidentService {
  private readonly http = inject(HttpClient);

  private readonly _incidents = signal<Incident[]>(this.generateMockIncidents());
  readonly incidents = this._incidents.asReadonly();

  private readonly _selectedIncident = signal<Incident | null>(null);
  readonly selectedIncident = this._selectedIncident.asReadonly();

  constructor() {
    this.fetchIncidents();
  }

  fetchIncidents(status?: IncidentStatus): void {
    let params = new HttpParams();
    if (status) {
      params = params.set('status', status);
    }

    this.http.get<any>(`${environment.apiBaseUrl}/incidents`, { params }).pipe(
      map((res) => {
        const rawList = res?.data?.content || res?.data || res?.content || (Array.isArray(res) ? res : []);
        if (!Array.isArray(rawList) || rawList.length === 0) {
          return this.generateMockIncidents();
        }
        return rawList.map(this.mapIncident);
      }),
      catchError(() => of(this.generateMockIncidents()))
    ).subscribe((list) => {
      this._incidents.set(list);
    });
  }

  fetchIncidentById(id: string): Observable<Incident> {
    return this.http.get<any>(`${environment.apiBaseUrl}/incidents/${id}`).pipe(
      map((res) => {
        const raw = res?.data || res;
        return this.mapIncident(raw);
      }),
      tap((inc) => this._selectedIncident.set(inc)),
      catchError(() => {
        const found = this._incidents().find((i) => i.id === id) || this.generateMockIncidents()[0];
        this._selectedIncident.set(found);
        return of(found);
      })
    );
  }

  assignAnalyst(id: string, assignedAnalyst: string): Observable<any> {
    return this.http.post<any>(`${environment.apiBaseUrl}/incidents/${id}/assign`, { assignedAnalyst, assignedBy: 'SOC_LEAD' }).pipe(
      tap(() => this.fetchIncidentById(id).subscribe()),
      catchError(() => {
        this._incidents.update((list) =>
          list.map((i) => (i.id === id ? { ...i, assignedAnalyst, status: 'ACKNOWLEDGED' } : i))
        );
        const sel = this._selectedIncident();
        if (sel && sel.id === id) {
          this._selectedIncident.set({ ...sel, assignedAnalyst, status: 'ACKNOWLEDGED' });
        }
        return of(null);
      })
    );
  }

  updateStatus(id: string, status: IncidentStatus, notes?: string): Observable<any> {
    return this.http.put<any>(`${environment.apiBaseUrl}/incidents/${id}/status`, { status, notes, updatedBy: 'SOC_ANALYST' }).pipe(
      tap(() => this.fetchIncidentById(id).subscribe()),
      catchError(() => {
        this._incidents.update((list) =>
          list.map((i) => (i.id === id ? { ...i, status } : i))
        );
        const sel = this._selectedIncident();
        if (sel && sel.id === id) {
          this._selectedIncident.set({ ...sel, status });
        }
        return of(null);
      })
    );
  }

  addNote(id: string, content: string): Observable<IncidentNote> {
    return this.http.post<any>(`${environment.apiBaseUrl}/incidents/${id}/notes`, { content, author: 'SOC_ANALYST' }).pipe(
      map((res) => res?.data || res),
      tap(() => this.fetchIncidentById(id).subscribe()),
      catchError(() => {
        const newNote: IncidentNote = {
          id: crypto.randomUUID(),
          incidentId: id,
          author: 'SOC_ANALYST',
          content,
          createdAt: new Date().toISOString()
        };
        const sel = this._selectedIncident();
        if (sel && sel.id === id) {
          this._selectedIncident.set({
            ...sel,
            notes: [newNote, ...(sel.notes || [])]
          });
        }
        return of(newNote);
      })
    );
  }

  performAction(id: string, actionType: IncidentActionType, target?: string, details?: string): Observable<any> {
    return this.http.post<any>(`${environment.apiBaseUrl}/incidents/${id}/actions`, {
      actionType,
      performedBy: 'SOC_ANALYST',
      target,
      details
    }).pipe(
      tap(() => this.fetchIncidentById(id).subscribe()),
      catchError(() => {
        const sel = this._selectedIncident();
        if (sel && sel.id === id) {
          const newStatus = actionType === 'RESOLVE' ? 'RESOLVED' : (actionType === 'ACKNOWLEDGE' ? 'ACKNOWLEDGED' : sel.status);
          this._selectedIncident.set({
            ...sel,
            status: newStatus,
            actions: [
              {
                id: crypto.randomUUID(),
                incidentId: id,
                actionType,
                performedBy: 'SOC_ANALYST',
                target: target || sel.affectedIp || sel.affectedUser,
                details: details || `Action executed: ${actionType}`,
                timestamp: new Date().toISOString()
              },
              ...(sel.actions || [])
            ]
          });
        }
        return of(null);
      })
    );
  }

  searchThreats(filter: ThreatHuntingFilter): Observable<ThreatHuntingResult[]> {
    let params = new HttpParams();
    if (filter.username) params = params.set('username', filter.username);
    if (filter.ipAddress) params = params.set('ipAddress', filter.ipAddress);
    if (filter.device) params = params.set('device', filter.device);
    if (filter.eventType) params = params.set('eventType', filter.eventType);
    if (filter.minRiskScore) params = params.set('minRiskScore', filter.minRiskScore.toString());

    return this.http.get<any>(`${environment.apiBaseUrl}/threat-hunting/search`, { params }).pipe(
      map((res) => res?.data || res || []),
      catchError(() => of(this.generateMockThreatResults()))
    );
  }

  triggerTestScenario(): Observable<any> {
    return this.http.post<any>(`${environment.apiBaseUrl}/incidents/test/trigger-impossible-travel`, {}).pipe(
      tap(() => this.fetchIncidents()),
      catchError(() => of(null))
    );
  }

  private mapIncident(a: any): Incident {
    return {
      id: a.id || crypto.randomUUID(),
      incidentNumber: a.incidentNumber || 'INC-2026-0001',
      alertId: a.alertId,
      title: a.title || 'Security Incident Detected',
      description: a.description || 'Behavioral anomaly triggered critical security threshold.',
      severity: a.severity || 'HIGH',
      status: a.status || 'OPEN',
      assignedAnalyst: a.assignedAnalyst,
      riskScore: a.riskScore || 88,
      aiConfidence: a.aiConfidence || 92.5,
      affectedUser: a.affectedUser || 'victim_user',
      affectedIp: a.affectedIp || '192.168.1.100',
      affectedDevice: a.affectedDevice || 'Linux x86_64 / Terminal',
      evidenceJson: a.evidenceJson,
      createdAt: a.createdAt || new Date().toISOString(),
      updatedAt: a.updatedAt || new Date().toISOString(),
      resolvedAt: a.resolvedAt,
      timeline: a.timeline || [],
      actions: a.actions || [],
      notes: a.notes || []
    };
  }

  private generateMockIncidents(): Incident[] {
    const now = new Date();
    return [
      {
        id: 'inc-001',
        incidentNumber: 'INC-2026-0001',
        alertId: 'alt-001',
        title: 'CRITICAL: Impossible Travel Attack Detected',
        description: 'User authenticated from New York, then 5 minutes later from Tokyo. AI Anomaly confidence 98.5%.',
        severity: 'CRITICAL',
        status: 'OPEN',
        assignedAnalyst: undefined,
        riskScore: 98,
        aiConfidence: 98.5,
        affectedUser: 'john_doe@sentinel.sec',
        affectedIp: '203.0.113.195',
        affectedDevice: 'MacBookPro18,1 / Safari 17.4',
        evidenceJson: '{"login1":"NYC","login2":"TYO","deltaMinutes":5}',
        createdAt: new Date(now.getTime() - 15 * 60000).toISOString(),
        updatedAt: new Date(now.getTime() - 15 * 60000).toISOString(),
        timeline: [
          {
            id: 't-1',
            incidentId: 'inc-001',
            eventType: 'LOGIN_FAILED',
            source: 'Auth Service',
            summary: 'Multiple failed password attempts for user john_doe@sentinel.sec',
            timestamp: new Date(now.getTime() - 20 * 60000).toISOString()
          },
          {
            id: 't-2',
            incidentId: 'inc-001',
            eventType: 'BRUTE_FORCE_DETECTED',
            source: 'Threat Engine',
            summary: 'Brute force pattern matched (12 attempts in 30s)',
            timestamp: new Date(now.getTime() - 18 * 60000).toISOString()
          },
          {
            id: 't-3',
            incidentId: 'inc-001',
            eventType: 'AI_ANOMALY',
            source: 'AI Behavioral Engine',
            summary: 'Impossible travel anomaly detected with 98.5% confidence',
            timestamp: new Date(now.getTime() - 16 * 60000).toISOString()
          },
          {
            id: 't-4',
            incidentId: 'inc-001',
            eventType: 'RISK_CRITICAL',
            source: 'Risk Engine',
            summary: 'Calculated risk score 98/100 exceeded threshold',
            timestamp: new Date(now.getTime() - 15 * 60000).toISOString()
          },
          {
            id: 't-5',
            incidentId: 'inc-001',
            eventType: 'ALERT_CREATED',
            source: 'Alert Engine',
            summary: 'Alert ALT-8921 created and escalated to SOC Command Center',
            timestamp: new Date(now.getTime() - 15 * 60000).toISOString()
          }
        ],
        actions: [],
        notes: [
          {
            id: 'n-1',
            incidentId: 'inc-001',
            author: 'SYSTEM',
            content: 'Incident automatically created from CRITICAL risk event.',
            createdAt: new Date(now.getTime() - 15 * 60000).toISOString()
          }
        ]
      },
      {
        id: 'inc-002',
        incidentNumber: 'INC-2026-0002',
        alertId: 'alt-002',
        title: 'HIGH: SQL Injection Exploit Attempt',
        description: 'Malicious SQL payload detected in authorization header query parameter.',
        severity: 'HIGH',
        status: 'INVESTIGATING',
        assignedAnalyst: 'sarah_connor',
        riskScore: 85,
        aiConfidence: 89.0,
        affectedUser: 'api_gateway_svc',
        affectedIp: '198.51.100.42',
        affectedDevice: 'curl/7.68.0',
        evidenceJson: '{"query":"SELECT * FROM users WHERE 1=1--"}',
        createdAt: new Date(now.getTime() - 45 * 60000).toISOString(),
        updatedAt: new Date(now.getTime() - 25 * 60000).toISOString(),
        timeline: [
          {
            id: 't-201',
            incidentId: 'inc-002',
            eventType: 'SQL_INJECTION_DETECTED',
            source: 'Threat Engine',
            summary: 'Pattern UNION SELECT matched in HTTP request body',
            timestamp: new Date(now.getTime() - 45 * 60000).toISOString()
          }
        ],
        actions: [
          {
            id: 'a-1',
            incidentId: 'inc-002',
            actionType: 'ASSIGN',
            performedBy: 'SOC_LEAD',
            target: 'sarah_connor',
            details: 'Assigned to Senior SOC Analyst Sarah',
            timestamp: new Date(now.getTime() - 30 * 60000).toISOString()
          }
        ],
        notes: [
          {
            id: 'n-2',
            incidentId: 'inc-002',
            author: 'sarah_connor',
            content: 'Analyzing ingress WAF rule block logs.',
            createdAt: new Date(now.getTime() - 25 * 60000).toISOString()
          }
        ]
      }
    ];
  }

  private generateMockThreatResults(): ThreatHuntingResult[] {
    return [
      {
        id: 'th-1',
        incidentNumber: 'INC-2026-0001',
        title: 'Impossible Travel Anomaly',
        eventType: 'IMPOSSIBLE_TRAVEL_ATTACK',
        source: 'AI Engine',
        user: 'john_doe@sentinel.sec',
        ipAddress: '203.0.113.195',
        device: 'MacBookPro / Safari',
        riskScore: 98,
        severity: 'CRITICAL',
        timestamp: new Date().toISOString(),
        incidentId: 'inc-001'
      },
      {
        id: 'th-2',
        incidentNumber: 'INC-2026-0002',
        title: 'SQL Injection Signature Match',
        eventType: 'SQL_INJECTION',
        source: 'Threat Engine',
        user: 'api_gateway_svc',
        ipAddress: '198.51.100.42',
        device: 'curl/7.68.0',
        riskScore: 85,
        severity: 'HIGH',
        timestamp: new Date(Date.now() - 3600000).toISOString(),
        incidentId: 'inc-002'
      }
    ];
  }
}

File: backend/alert-service/src/main/java/com/sentinel/alert/incident/service/IncidentService.java
package com.sentinel.alert.incident.service;

import com.sentinel.alert.domain.entity.AlertEntity;
import com.sentinel.alert.domain.model.AlertSeverity;
import com.sentinel.alert.incident.domain.entity.IncidentActionEntity;
import com.sentinel.alert.incident.domain.entity.IncidentEntity;
import com.sentinel.alert.incident.domain.entity.IncidentNoteEntity;
import com.sentinel.alert.incident.domain.entity.IncidentTimelineEventEntity;
import com.sentinel.alert.incident.domain.model.IncidentActionType;
import com.sentinel.alert.incident.domain.model.IncidentStatus;
import com.sentinel.alert.incident.dto.AddIncidentNoteRequest;
import com.sentinel.alert.incident.dto.AssignIncidentRequest;
import com.sentinel.alert.incident.dto.IncidentActionResponse;
import com.sentinel.alert.incident.dto.IncidentNoteResponse;
import com.sentinel.alert.incident.dto.IncidentResponse;
import com.sentinel.alert.incident.dto.IncidentTimelineEventResponse;
import com.sentinel.alert.incident.dto.PerformIncidentActionRequest;
import com.sentinel.alert.incident.dto.UpdateIncidentStatusRequest;
import com.sentinel.alert.incident.repository.IncidentActionRepository;
import com.sentinel.alert.incident.repository.IncidentNoteRepository;
import com.sentinel.alert.incident.repository.IncidentRepository;
import com.sentinel.alert.incident.repository.IncidentTimelineEventRepository;
import com.sentinel.common.exception.BusinessException;
import com.sentinel.common.exception.ErrorCode;
import java.time.Instant;
import java.util.List;
import java.util.UUID;
import java.util.concurrent.atomic.AtomicLong;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

@Service
public class IncidentService {
    private static final Logger log = LoggerFactory.getLogger(IncidentService.class);
    private static final AtomicLong INCIDENT_COUNTER = new AtomicLong(1000);

    private final IncidentRepository incidentRepository;
    private final IncidentTimelineEventRepository timelineRepository;
    private final IncidentActionRepository actionRepository;
    private final IncidentNoteRepository noteRepository;

    public IncidentService(IncidentRepository incidentRepository,
                           IncidentTimelineEventRepository timelineRepository,
                           IncidentActionRepository actionRepository,
                           IncidentNoteRepository noteRepository) {
        this.incidentRepository = incidentRepository;
        this.timelineRepository = timelineRepository;
        this.actionRepository = actionRepository;
        this.noteRepository = noteRepository;
    }

    @Transactional
    public IncidentEntity createIncidentFromAlert(AlertEntity alert) {
        if (alert == null) {
            return null;
        }

        // Prevent duplicate creation for the same alert
        if (alert.getId() != null && incidentRepository.existsByAlertId(alert.getId())) {
            log.info("Incident already exists for alertId={}", alert.getId());
            return incidentRepository.findByAlertId(alert.getId()).orElse(null);
        }

        String incidentNum = generateIncidentNumber();
        double aiConfidence = alert.getRiskScore() > 80 ? 94.5 : (alert.getRiskScore() > 50 ? 82.0 : 65.0);

        String user = alert.getAffectedUser() != null ? alert.getAffectedUser() : "unknown_user";
        String ip = alert.getAffectedIp() != null ? alert.getAffectedIp() : "192.168.1.100";
        String device = "Linux x86_64 / SOC Terminal";

        IncidentEntity incident = new IncidentEntity(
                incidentNum,
                alert.getId(),
                alert.getTitle(),
                alert.getDescription(),
                alert.getSeverity(),
                IncidentStatus.OPEN,
                alert.getAssignedAnalyst(),
                alert.getRiskScore(),
                aiConfidence,
                user,
                ip,
                device,
                alert.getEvidenceJson()
        );

        IncidentEntity saved = incidentRepository.save(incident);
        log.info("INCIDENT_CREATED incidentNumber={} alertId={} severity={} riskScore={}",
                saved.getIncidentNumber(), saved.getAlertId(), saved.getSeverity(), saved.getRiskScore());

        // Build initial attack timeline based on evidence & alert attributes
        buildInitialTimeline(saved, alert);

        return saved;
    }

    private void buildInitialTimeline(IncidentEntity incident, AlertEntity alert) {
        Instant baseTime = alert.getCreatedAt() != null ? alert.getCreatedAt() : Instant.now();

        // 1. Initial Access Vector / Event
        timelineRepository.save(new IncidentTimelineEventEntity(
                incident.getId(),
                "LOGIN_FAILED",
                "Authentication Gateway",
                "Multiple failed authentication attempts detected for user: " + incident.getAffectedUser(),
                "{\"user\":\"" + incident.getAffectedUser() + "\",\"ip\":\"" + incident.getAffectedIp() + "\"}",
                baseTime.minusSeconds(180)
        ));

        // 2. Threat Detection
        String threatTitle = alert.getAlertType() != null ? alert.getAlertType() : "BRUTE_FORCE_DETECTED";
        timelineRepository.save(new IncidentTimelineEventEntity(
                incident.getId(),
                threatTitle,
                "Threat Detection Engine",
                "Attack signature matched: " + threatTitle + " on endpoint " + (alert.getAffectedApi() != null ? alert.getAffectedApi() : "/api/v1/auth/login"),
                "{\"signature\":\"SIG-RULE-" + (Math.abs(incident.getIncidentNumber().hashCode() % 900) + 100) + "\"}",
                baseTime.minusSeconds(120)
        ));

        // 3. AI Anomaly Prediction
        timelineRepository.save(new IncidentTimelineEventEntity(
                incident.getId(),
                "AI_ANOMALY",
                "AI Behavioral Engine",
                "Anomalous access pattern detected with " + incident.getAiConfidence() + "% confidence score",
                "{\"aiConfidence\":" + incident.getAiConfidence() + ",\"anomalyType\":\"BEHAVIORAL_DEVIATION\"}",
                baseTime.minusSeconds(60)
        ));

        // 4. Risk Engine Decision
        timelineRepository.save(new IncidentTimelineEventEntity(
                incident.getId(),
                "RISK_CRITICAL",
                "Risk Engine",
                "Calculated risk score " + incident.getRiskScore() + "/100 — threshold exceeded for " + incident.getSeverity() + " alert level",
                "{\"riskScore\":" + incident.getRiskScore() + ",\"severity\":\"" + incident.getSeverity() + "\"}",
                baseTime.minusSeconds(30)
        ));

        // 5. Alert Created
        timelineRepository.save(new IncidentTimelineEventEntity(
                incident.getId(),
                "ALERT_CREATED",
                "Alert Management Engine",
                "Alert " + alert.getAlertCode() + " created and dispatched to SOC Command Center",
                "{\"alertCode\":\"" + alert.getAlertCode() + "\",\"incidentNumber\":\"" + incident.getIncidentNumber() + "\"}",
                baseTime
        ));
    }

    @Transactional(readOnly = true)
    public Page<IncidentResponse> getIncidents(IncidentStatus status, AlertSeverity severity, Pageable pageable) {
        if (status != null) {
            return incidentRepository.findByStatus(status, pageable).map(this::toIncidentResponseSummary);
        }
        if (severity != null) {
            return incidentRepository.findBySeverity(severity, pageable).map(this::toIncidentResponseSummary);
        }
        return incidentRepository.findAll(pageable).map(this::toIncidentResponseSummary);
    }

    @Transactional(readOnly = true)
    public IncidentResponse getIncidentById(UUID id) {
        IncidentEntity incident = incidentRepository.findById(id)
                .orElseThrow(() -> new BusinessException(ErrorCode.RESOURCE_NOT_FOUND, "Incident not found: " + id));
        return toFullIncidentResponse(incident);
    }

    @Transactional
    public IncidentResponse assignIncident(UUID id, AssignIncidentRequest request) {
        IncidentEntity incident = incidentRepository.findById(id)
                .orElseThrow(() -> new BusinessException(ErrorCode.RESOURCE_NOT_FOUND, "Incident not found: " + id));

        String analyst = request.assignedAnalyst();
        incident.assignAnalyst(analyst);
        IncidentEntity saved = incidentRepository.save(incident);

        String performedBy = request.assignedBy() != null ? request.assignedBy() : "SOC_LEAD";
        actionRepository.save(new IncidentActionEntity(
                saved.getId(),
                IncidentActionType.ASSIGN,
                performedBy,
                analyst,
                "Assigned incident " + saved.getIncidentNumber() + " to analyst " + analyst
        ));

        timelineRepository.save(new IncidentTimelineEventEntity(
                saved.getId(),
                "ANALYST_ASSIGNED",
                "SOC Command Center",
                "Incident assigned to analyst: " + analyst,
                "{\"assignedBy\":\"" + performedBy + "\",\"analyst\":\"" + analyst + "\"}",
                Instant.now()
        ));

        return toFullIncidentResponse(saved);
    }

    @Transactional
    public IncidentResponse updateIncidentStatus(UUID id, UpdateIncidentStatusRequest request) {
        IncidentEntity incident = incidentRepository.findById(id)
                .orElseThrow(() -> new BusinessException(ErrorCode.RESOURCE_NOT_FOUND, "Incident not found: " + id));

        validateStatusTransition(incident.getStatus(), request.status());

        IncidentStatus prevStatus = incident.getStatus();
        incident.updateStatus(request.status());
        IncidentEntity saved = incidentRepository.save(incident);

        String performedBy = request.updatedBy() != null ? request.updatedBy() : "SOC_ANALYST";
        actionRepository.save(new IncidentActionEntity(
                saved.getId(),
                mapStatusToActionType(request.status()),
                performedBy,
                saved.getIncidentNumber(),
                "Status updated from " + prevStatus + " to " + request.status() + ". Notes: " + (request.notes() != null ? request.notes() : "N/A")
        ));

        timelineRepository.save(new IncidentTimelineEventEntity(
                saved.getId(),
                "STATUS_UPDATED_" + request.status(),
                "SOC Command Center",
                "Incident status changed from " + prevStatus + " to " + request.status(),
                "{\"previousStatus\":\"" + prevStatus + "\",\"newStatus\":\"" + request.status() + "\"}",
                Instant.now()
        ));

        if (request.notes() != null && !request.notes().isBlank()) {
            noteRepository.save(new IncidentNoteEntity(saved.getId(), performedBy, request.notes()));
        }

        return toFullIncidentResponse(saved);
    }

    @Transactional
    public IncidentNoteResponse addNote(UUID id, AddIncidentNoteRequest request) {
        IncidentEntity incident = incidentRepository.findById(id)
                .orElseThrow(() -> new BusinessException(ErrorCode.RESOURCE_NOT_FOUND, "Incident not found: " + id));

        String author = request.author() != null ? request.author() : "SOC_ANALYST";
        IncidentNoteEntity note = new IncidentNoteEntity(incident.getId(), author, request.content());
        IncidentNoteEntity saved = noteRepository.save(note);

        timelineRepository.save(new IncidentTimelineEventEntity(
                incident.getId(),
                "NOTE_ADDED",
                "Analyst Workbench",
                "Note added by " + author,
                "{\"author\":\"" + author + "\",\"noteId\":\"" + saved.getId() + "\"}",
                Instant.now()
        ));

        return new IncidentNoteResponse(saved.getId(), saved.getIncidentId(), saved.getAuthor(), saved.getContent(), saved.getCreatedAt());
    }

    @Transactional
    public IncidentResponse performAction(UUID id, PerformIncidentActionRequest request) {
        IncidentEntity incident = incidentRepository.findById(id)
                .orElseThrow(() -> new BusinessException(ErrorCode.RESOURCE_NOT_FOUND, "Incident not found: " + id));

        String performedBy = request.performedBy() != null ? request.performedBy() : "SOC_ANALYST";
        String target = request.target() != null ? request.target() : getActionDefaultTarget(incident, request.actionType());
        String actionDetails = request.details();

        String auditDetails;
        switch (request.actionType()) {
            case ACKNOWLEDGE -> {
                incident.updateStatus(IncidentStatus.ACKNOWLEDGED);
                auditDetails = "Incident acknowledged by SOC analyst";
            }
            case ASSIGN -> {
                incident.assignAnalyst(target);
                auditDetails = "Assigned incident to " + target;
            }
            case BLOCK_IP -> {
                auditDetails = "Action recorded — enforcement integration pending (IP: " + target + "). " + (actionDetails != null ? actionDetails : "");
            }
            case DISABLE_USER -> {
                auditDetails = "Action recorded — enforcement integration pending (User: " + target + "). " + (actionDetails != null ? actionDetails : "");
            }
            case RESET_PASSWORD -> {
                auditDetails = "Action recorded — enforcement integration pending (User: " + target + "). " + (actionDetails != null ? actionDetails : "");
            }
            case RESOLVE -> {
                incident.updateStatus(IncidentStatus.RESOLVED);
                auditDetails = "Incident resolved. Details: " + (actionDetails != null ? actionDetails : "Mitigation verified");
            }
            default -> auditDetails = "Action executed: " + request.actionType();
        }

        IncidentEntity saved = incidentRepository.save(incident);

        actionRepository.save(new IncidentActionEntity(
                saved.getId(),
                request.actionType(),
                performedBy,
                target,
                auditDetails
        ));

        timelineRepository.save(new IncidentTimelineEventEntity(
                saved.getId(),
                "ACTION_" + request.actionType().name(),
                "Analyst Actions",
                "Analyst action executed: " + request.actionType().name() + " on target " + target,
                "{\"actionType\":\"" + request.actionType().name() + "\",\"performedBy\":\"" + performedBy + "\",\"details\":\"" + auditDetails + "\"}",
                Instant.now()
        ));

        return toFullIncidentResponse(saved);
    }

    private void validateStatusTransition(IncidentStatus current, IncidentStatus target) {
        if (current == target) {
            return;
        }
        boolean valid = switch (current) {
            case OPEN -> target == IncidentStatus.ACKNOWLEDGED || target == IncidentStatus.INVESTIGATING || target == IncidentStatus.RESOLVED;
            case ACKNOWLEDGED -> target == IncidentStatus.INVESTIGATING || target == IncidentStatus.MITIGATED || target == IncidentStatus.RESOLVED;
            case INVESTIGATING -> target == IncidentStatus.MITIGATED || target == IncidentStatus.RESOLVED;
            case MITIGATED -> target == IncidentStatus.RESOLVED || target == IncidentStatus.CLOSED;
            case RESOLVED -> target == IncidentStatus.CLOSED || target == IncidentStatus.INVESTIGATING;
            case CLOSED -> false;
        };

        if (!valid) {
            throw new BusinessException(ErrorCode.VALIDATION_FAILED, "Invalid status transition from " + current + " to " + target);
        }
    }

    private String getActionDefaultTarget(IncidentEntity incident, IncidentActionType actionType) {
        return switch (actionType) {
            case BLOCK_IP -> incident.getAffectedIp() != null ? incident.getAffectedIp() : "192.168.1.100";
            case DISABLE_USER, RESET_PASSWORD -> incident.getAffectedUser() != null ? incident.getAffectedUser() : "affected_user";
            case ASSIGN -> incident.getAssignedAnalyst() != null ? incident.getAssignedAnalyst() : "soc_analyst";
            default -> incident.getIncidentNumber();
        };
    }

    private IncidentActionType mapStatusToActionType(IncidentStatus status) {
        return switch (status) {
            case ACKNOWLEDGED -> IncidentActionType.ACKNOWLEDGE;
            case RESOLVED -> IncidentActionType.RESOLVE;
            default -> IncidentActionType.ASSIGN;
        };
    }

    private String generateIncidentNumber() {
        return "INC-2026-" + String.format("%04d", INCIDENT_COUNTER.getAndIncrement());
    }

    private IncidentResponse toIncidentResponseSummary(IncidentEntity entity) {
        return new IncidentResponse(
                entity.getId(),
                entity.getIncidentNumber(),
                entity.getAlertId(),
                entity.getTitle(),
                entity.getDescription(),
                entity.getSeverity(),
                entity.getStatus(),
                entity.getAssignedAnalyst(),
                entity.getRiskScore(),
                entity.getAiConfidence(),
                entity.getAffectedUser(),
                entity.getAffectedIp(),
                entity.getAffectedDevice(),
                entity.getEvidenceJson(),
                entity.getCreatedAt(),
                entity.getUpdatedAt(),
                entity.getResolvedAt(),
                List.of(),
                List.of(),
                List.of()
        );
    }

    private IncidentResponse toFullIncidentResponse(IncidentEntity entity) {
        List<IncidentTimelineEventResponse> timeline = timelineRepository.findByIncidentIdOrderByTimestampAsc(entity.getId()).stream()
                .map(t -> new IncidentTimelineEventResponse(t.getId(), t.getIncidentId(), t.getEventType(), t.getSource(), t.getSummary(), t.getDetailsJson(), t.getTimestamp()))
                .toList();

        List<IncidentActionResponse> actions = actionRepository.findByIncidentIdOrderByTimestampDesc(entity.getId()).stream()
                .map(a -> new IncidentActionResponse(a.getId(), a.getIncidentId(), a.getActionType(), a.getPerformedBy(), a.getTarget(), a.getDetails(), a.getTimestamp()))
                .toList();

        List<IncidentNoteResponse> notes = noteRepository.findByIncidentIdOrderByCreatedAtDesc(entity.getId()).stream()
                .map(n -> new IncidentNoteResponse(n.getId(), n.getIncidentId(), n.getAuthor(), n.getContent(), n.getCreatedAt()))
                .toList();

        return new IncidentResponse(
                entity.getId(),
                entity.getIncidentNumber(),
                entity.getAlertId(),
                entity.getTitle(),
                entity.getDescription(),
                entity.getSeverity(),
                entity.getStatus(),
                entity.getAssignedAnalyst(),
                entity.getRiskScore(),
                entity.getAiConfidence(),
                entity.getAffectedUser(),
                entity.getAffectedIp(),
                entity.getAffectedDevice(),
                entity.getEvidenceJson(),
                entity.getCreatedAt(),
                entity.getUpdatedAt(),
                entity.getResolvedAt(),
                timeline,
                actions,
                notes
        );
    }
}
